File: src/lib/worktreeManager.ts
import { invoke } from "@tauri-apps/api/core";
import { homeDir } from "@tauri-apps/api/path";

/** Worktree info from the backend. */
export interface WorktreeInfo {
  path: string;
  head: string;
  branch: string | null;
  is_bare: boolean;
  /** True for the main working tree (the original clone directory). */
  is_main_worktree: boolean;
}

/** Result of preparing a worktree for a session. */
export interface WorktreePreparationResult {
  /** The directory where the session should run (worktree or project path). */
  working_directory: string;
  /** The worktree path if one was created or reused. */
  worktree_path: string | null;
  /** The resolved branch name (auto-detected or explicitly specified). */
  branch: string | null;
  /** Whether a new worktree was created (vs. reused or skipped). */
  created: boolean;
  /** Warning message if something unexpected happened but we recovered. */
  warning: string | null;
}

/**
 * Generates a hash from a string for creating unique worktree paths.
 */
function hashString(str: string): string {
  let hash = 0;
  for (let i = 0; i < str.length; i++) {
    const char = str.charCodeAt(i);
    hash = ((hash << 5) - hash) + char;
    hash = hash & hash; // Convert to 32bit integer
  }
  return Math.abs(hash).toString(16).slice(0, 8);
}

/**
 * Sanitizes a branch name for use in filesystem paths.
 */
function sanitizeBranch(branch: string): string {
  return branch.replace(/[^a-zA-Z0-9_-]/g, "-");
}

/**
 * Gets the base directory for worktrees.
 * Uses ~/.claude-maestro/worktrees/
 */
async function getWorktreeBaseDir(): Promise<string> {
  const home = await homeDir();
  return `${home}.claude-maestro/worktrees`;
}

/**
 * Calculates the worktree path for a given repo and branch.
 *
 * Path format: ~/.claude-maestro/worktrees/{repoHash}/{sanitizedBranch}/
 *
 * @param repoPath - The path to the main repository
 * @param branch - The branch name
 * @returns The worktree path
 */
export async function getWorktreePath(repoPath: string, branch: string): Promise<string> {
  const baseDir = await getWorktreeBaseDir();
  const repoHash = hashString(repoPath);
  const sanitizedBranch = sanitizeBranch(branch);
  return `${baseDir}/${repoHash}/${sanitizedBranch}`;
}

/**
 * Creates a worktree for a session on a specific branch.
 *
 * If the worktree already exists for this branch, returns its path.
 * If a new branch is needed, creates it from the current HEAD.
 *
 * @param repoPath - The path to the main repository
 * @param sessionId - The session ID (for logging)
 * @param branch - The branch to checkout in the worktree
 * @param createBranch - Whether to create a new branch (default: false)
 * @returns The worktree path
 */
export async function createSessionWorktree(
  repoPath: string,
  sessionId: number,
  branch: string,
  createBranch = false
): Promise<string> {
  const worktreePath = await getWorktreePath(repoPath, branch);

  try {
    // Check if worktree already exists
    const existingWorktrees = await invoke<WorktreeInfo[]>("git_worktree_list", {
      repoPath,
    });

    const existing = existingWorktrees.find((wt) => wt.path === worktreePath);
    if (existing) {
      console.log(`[Session ${sessionId}] Worktree already exists at ${worktreePath}`);
      return worktreePath;
    }

    // Create the worktree
    const result = await invoke<WorktreeInfo>("git_worktree_add", {
      repoPath,
      path: worktreePath,
      newBranch: createBranch ? branch : null,
      checkoutRef: createBranch ? null : branch,
    });

    console.log(`[Session ${sessionId}] Created worktree at ${result.path} on branch ${result.branch}`);
    return result.path;
  } catch (err) {
    console.error(`[Session ${sessionId}] Failed to create worktree:`, err);
    throw err;
  }
}

/**
 * Removes a worktree associated with a session.
 *
 * @param repoPath - The path to the main repository
 * @param worktreePath - The worktree path to remove
 * @param force - Whether to force removal even with uncommitted changes
 */
export async function removeSessionWorktree(
  repoPath: string,
  worktreePath: string,
  force = false
): Promise<void> {
  try {
    await invoke("git_worktree_remove", {
      repoPath,
      path: worktreePath,
      force,
    });
    console.log(`Removed worktree at ${worktreePath}`);
  } catch (err) {
    console.error(`Failed to remove worktree at ${worktreePath}:`, err);
    throw err;
  }
}

/**
 * Lists all worktrees for a repository.
 *
 * @param repoPath - The path to the main repository
 * @returns List of worktree info
 */
export async function listWorktrees(repoPath: string): Promise<WorktreeInfo[]> {
  return invoke<WorktreeInfo[]>("git_worktree_list", { repoPath });
}

/**
 * Checks if a worktree exists for a given branch.
 *
 * @param repoPath - The path to the main repository
 * @param branch - The branch name to check
 * @returns True if a worktree exists for this branch
 */
export async function worktreeExistsForBranch(
  repoPath: string,
  branch: string
): Promise<boolean> {
  const worktrees = await listWorktrees(repoPath);
  return worktrees.some((wt) => wt.branch === branch);
}

/**
 * Gets the worktree info for a specific branch if it exists.
 *
 * @param repoPath - The path to the main repository
 * @param branch - The branch name
 * @returns Worktree info or null if not found
 */
export async function getWorktreeForBranch(
  repoPath: string,
  branch: string
): Promise<WorktreeInfo | null> {
  const worktrees = await listWorktrees(repoPath);
  return worktrees.find((wt) => wt.branch === branch) ?? null;
}

/**
 * Prepares a worktree for a session, handling all edge cases gracefully.
 *
 * This function orchestrates worktree creation for a session launch:
 * - If no branch is specified, returns the project path as-is.
 * - If a worktree already exists for this branch, reuses it.
 * - If the branch is checked out in the main repo, switches main to default branch first.
 * - If the branch doesn't exist locally, creates it from HEAD.
 * - Creates the worktree via the backend WorktreeManager.
 *
 * On any failure, falls back to the project path so sessions always launch.
 *
 * @param projectPath - The path to the main repository
 * @param branch - The branch to checkout in the worktree (null to skip worktree)
 * @returns The preparation result with the working directory to use
 */
export async function prepareSessionWorktree(
  projectPath: string,
  branch: string | null,
  worktreeBasePath?: string | null,
  forceNew?: boolean,
): Promise<WorktreePreparationResult> {
  try {
    const result = await invoke<WorktreePreparationResult>("prepare_session_worktree", {
      projectPath,
      branch,
      worktreeBasePath: worktreeBasePath ?? null,
      forceNew: forceNew ?? false,
    });

    if (result.warning) {
      console.warn(`Worktree warning: ${result.warning}`);
    }

    if (result.created) {
      console.log(`Created worktree at ${result.worktree_path}`);
    } else if (result.worktree_path) {
      console.log(`Reusing worktree at ${result.worktree_path}`);
    } else {
      console.log(`Using project path (no worktree)`);
    }

    return result;
  } catch (err) {
    console.error(`Failed to prepare worktree:`, err);
    // Fall back to project path on error
    return {
      working_directory: projectPath,
      worktree_path: null,
      branch: null,
      created: false,
      warning: `Failed to prepare worktree: ${err}`,
    };
  }
}

/**
 * Cleans up a worktree when a session ends.
 *
 * Removes the worktree from the filesystem and prunes git refs.
 * Failures are logged but don't prevent session cleanup.
 *
 * @param projectPath - The path to the main repository
 * @param worktreePath - The worktree path to clean up
 * @returns True if a worktree was cleaned up, false otherwise
 */
export async function cleanupSessionWorktree(
  projectPath: string,
  worktreePath: string | null
): Promise<boolean> {
  if (!worktreePath) {
    return false;
  }

  try {
    const cleaned = await invoke<boolean>("cleanup_session_worktree", {
      projectPath,
      worktreePath,
    });

    if (cleaned) {
      console.log(`Cleaned up worktree at ${worktreePath}`);
    }

    return cleaned;
  } catch (err) {
    console.error(`Failed to cleanup worktree at ${worktreePath}:`, err);
    return false;
  }
}

File: src-tauri/src/core/worktree_manager.rs
use std::collections::HashSet;
use std::path::{Path, PathBuf};

use sha2::{Digest, Sha256};

use crate::git::{Git, GitError, WorktreeInfo};

pub(crate) fn worktree_base_dir() -> PathBuf {
    directories::ProjectDirs::from("com", "maestro", "maestro")
        .map(|p| p.data_dir().to_path_buf())
        .unwrap_or_else(|| {
            std::env::var("HOME")
                .map(PathBuf::from)
                .map(|p| p.join(".local").join("share").join("maestro"))
                .expect("HOME environment variable must be set for worktree management")
        })
        .join("worktrees")
}

/// Extracts the project name from a repo path (last path component, lowercased).
/// Falls back to "project" if the path has no file name component.
fn project_name(repo_path: &Path) -> String {
    repo_path
        .file_name()
        .and_then(|n| n.to_str())
        .unwrap_or("project")
        .to_lowercase()
}


/// Replaces filesystem-unsafe characters in branch names with hyphens.
/// Covers `/`, `\`, `:`, `*`, `?`, `"`, `<`, `>`, and `|`.
/// Also handles `.` and `..` as special cases returning `unnamed-branch`.
fn sanitize_branch(branch: &str) -> String {
    if branch.is_empty() || branch == "." || branch == ".." {
        return "unnamed-branch".to_string();
    }

    let sanitized: String = branch
        .chars()
        .map(|c| match c {
            '/' | '\\' | ':' | '*' | '?' | '"' | '<' | '>' | '|' => '-',
            _ => c,
        })
        .collect();

    sanitized
}

/// Returns the override path if provided, otherwise falls back to the default
/// XDG-based worktree base directory.
fn effective_base_dir(base_override: Option<&Path>) -> PathBuf {
    base_override
        .map(|p| p.to_path_buf())
        .unwrap_or_else(worktree_base_dir)
}

/// Manages Maestro-owned git worktrees under a deterministic, repo-specific
/// directory inside XDG data dirs.
///
/// Worktree paths are derived from a SHA-256 hash of the canonical repo path
/// (truncated to 16 hex chars) so that different repos never collide, and a
/// sanitized branch name so each branch gets its own subdirectory.
pub struct WorktreeManager;

impl Default for WorktreeManager {
    fn default() -> Self {
        Self::new()
    }
}

impl WorktreeManager {
    /// Creates a new stateless manager. All path computation is pure and
    /// deterministic from the repo path and branch name.
    pub fn new() -> Self {
        Self
    }

    /// Compute the worktree path for a given repo + branch.
    pub(crate) async fn worktree_path(&self, repo_path: &Path, branch: &str) -> PathBuf {
        self.worktree_path_with_base(repo_path, branch, None).await
    }

    /// Compute the worktree path with an optional base directory override.
    pub(crate) async fn worktree_path_with_base(
        &self,
        repo_path: &Path,
        branch: &str,
        base_override: Option<&Path>,
    ) -> PathBuf {
        let name = project_name(repo_path);
        // Hash the branch name for a short, unique worktree identifier.
        // Pattern: <base>/<repoName>/moist-<8-char-hash>  (e.g. perfectbooth/moist-8477cc01)
        let digest = Sha256::digest(branch.as_bytes());
        let hash = &format!("{:x}", digest)[..8];
        let worktree_name = format!("moist-{}", hash);
        effective_base_dir(base_override).join(name).join(worktree_name)
    }

    /// Creates a worktree for the given branch, returning its path on disk.
    ///
    /// Checks that the branch is not already checked out in another worktree
    /// before creating (returns `BranchAlreadyCheckedOut` if so). Parent
    /// directories are created automatically. The worktree checks out the
    /// existing branch -- no new branch is created.
    pub async fn create(
        &self,
        branch: &str,
        repo_path: &Path,
    ) -> Result<PathBuf, GitError> {
        self.create_with_base(branch, repo_path, None, false).await
    }

    /// Creates a worktree with an optional base directory override.
    ///
    /// `force` — use `--force` on `git worktree add` (needed when the branch is
    /// already checked out in the main repo or another worktree).
    /// `force_new` — skip the duplicate-branch guard and append a unique suffix to the
    /// path so a fresh worktree is always created even if one already exists for this branch.
    pub async fn create_with_base(
        &self,
        branch: &str,
        repo_path: &Path,
        base_override: Option<&Path>,
        force: bool,
    ) -> Result<PathBuf, GitError> {
        self.create_with_base_inner(branch, repo_path, base_override, force, false).await
    }

    pub async fn create_with_base_new(
        &self,
        branch: &str,
        repo_path: &Path,
        base_override: Option<&Path>,
    ) -> Result<PathBuf, GitError> {
        self.create_with_base_inner(branch, repo_path, base_override, true, true).await
    }

    async fn create_with_base_inner(
        &self,
        branch: &str,
        repo_path: &Path,
        base_override: Option<&Path>,
        force: bool,
        force_new: bool,
    ) -> Result<PathBuf, GitError> {
        let git = Git::new(repo_path);

        let wt_path = self.worktree_path_with_base(repo_path, branch, base_override).await;

        // Guard against duplicate branch in non-main worktrees unless force_new.
        if !force_new {
            let existing = git.worktree_list().await?;
            for wt in &existing {
                if wt.is_main_worktree {
                    continue;
                }
                if let Some(ref wt_branch) = wt.branch {
                    if wt_branch == branch {
                        return Err(GitError::BranchAlreadyCheckedOut {
                            branch: branch.to_string(),
                            path: wt.path.clone(),
                        });
                    }
                }
            }
        } else {
            // force_new: tear down any existing worktree at this path (git-registered or not)
            // so git worktree add always gets a clean, non-existent target directory.
            if let Ok(existing) = git.worktree_list().await {
                for wt in &existing {
                    if wt.is_main_worktree {
                        continue;
                    }
                    if Path::new(&wt.path) == wt_path {
                        log::info!("force_new: unregistering existing worktree at {}", wt.path);
                        let _ = git.worktree_remove(Path::new(&wt.path), true).await;
                        let _ = git.worktree_prune().await;
                        break;
                    }
                }
            }
            // Remove the directory regardless of whether git knew about it —
            // git worktree add fails if the target path already exists on disk.
            if wt_path.exists() {
                log::info!("force_new: removing leftover directory at {}", wt_path.display());
                let _ = tokio::fs::remove_dir_all(&wt_path).await;
            }
        }

        // Create parent directories
        if let Some(parent) = wt_path.parent() {
            tokio::fs::create_dir_all(parent).await.map_err(|e| GitError::SpawnError {
                source: e,
                command: format!("create_dir_all {:?}", parent),
            })?;
        }

        if force || force_new {
            git.worktree_add_force(&wt_path, Some(branch)).await?;
        } else {
            git.worktree_add(&wt_path, None, Some(branch)).await?;
        }

        Ok(wt_path)
    }

    /// Force-removes a worktree and prunes its git ref, then attempts to
    /// clean up the empty parent directory (silently ignored if non-empty).
    pub async fn remove(&self, repo_path: &Path, wt_path: &Path) -> Result<(), GitError> {
        let git = Git::new(repo_path);
        git.worktree_remove(wt_path, true).await?;
        git.worktree_prune().await?;

        // Clean up empty parent directories
        if let Some(parent) = wt_path.parent() {
            let _ = tokio::fs::remove_dir(parent).await; // only succeeds if empty
        }

        Ok(())
    }

    /// Lists only worktrees that live under Maestro's managed base directory,
    /// filtering out the main worktree and any manually created worktrees.
    pub async fn list_managed(&self, repo_path: &Path) -> Result<Vec<WorktreeInfo>, GitError> {
        self.list_managed_with_base(repo_path, None).await
    }

    /// Lists managed worktrees with an optional base directory override.
    pub async fn list_managed_with_base(
        &self,
        repo_path: &Path,
        base_override: Option<&Path>,
    ) -> Result<Vec<WorktreeInfo>, GitError> {
        let git = Git::new(repo_path);
        let all = git.worktree_list().await?;

        let base = effective_base_dir(base_override);

        Ok(all
            .into_iter()
            .filter(|wt| Path::new(&wt.path).starts_with(&base))
            .collect())
    }

    /// Prunes stale git worktree refs and removes orphaned directories.
    ///
    /// First runs `git worktree prune`, then scans the managed directory for
    /// subdirectories that are no longer in git's worktree list. Orphaned
    /// directories are deleted with `remove_dir_all`. No-ops gracefully if
    /// the managed directory does not exist yet.
    pub async fn prune(&self, repo_path: &Path) -> Result<(), GitError> {
        let git = Git::new(repo_path);
        git.worktree_prune().await?;

        // Scan the project's managed dir for orphaned branch worktrees.
        // Pattern: <base>/<repoName>/<branch> — prune scans <base>/<repoName>/.
        let name = project_name(repo_path);
        let managed_dir = worktree_base_dir().join(&name);

        let base_exists = tokio::fs::try_exists(&managed_dir)
            .await
            .map_err(|e| GitError::SpawnError {
                source: e,
                command: format!("try_exists {:?}", managed_dir),
            })?;
        if !base_exists {
            return Ok(());
        }

        let active_raw: Vec<String> = git
            .worktree_list()
            .await?
            .iter()
            .map(|wt| wt.path.clone())
            .collect();

        // Canonicalize active paths for reliable comparison; fall back to raw path
        let mut active: HashSet<String> = HashSet::with_capacity(active_raw.len());
        for raw in &active_raw {
            let p = Path::new(raw);
            let canonical = tokio::fs::canonicalize(p).await.unwrap_or_else(|_| p.to_path_buf());
            active.insert(canonical.to_string_lossy().to_string());
        }

        if let Ok(mut entries) = tokio::fs::read_dir(&managed_dir).await {
            while let Ok(Some(entry)) = entries.next_entry().await {
                let path = entry.path();
                let canonical_entry = tokio::fs::canonicalize(&path)
                    .await
                    .unwrap_or_else(|_| path.clone());
                let entry_key = canonical_entry.to_string_lossy().to_string();
                let is_dir = tokio::fs::metadata(&path)
                    .await
                    .map(|m| m.is_dir())
                    .unwrap_or(false);
                if !active.contains(&entry_key) && is_dir {
                    log::info!("Removing orphaned worktree dir: {}", path.display());
                    let _ = tokio::fs::remove_dir_all(&path).await;
                }
            }
        }

        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::git::Git;
    use tempfile::tempdir;

    /// Helper: creates a temp git repo with an initial commit and returns its path.
    async fn create_test_repo() -> (tempfile::TempDir, PathBuf) {
        let dir = tempdir().unwrap();
        let path = dir.path().to_path_buf();
        let git = Git::new(&path);

        git.run(&["init"]).await.unwrap();
        git.run(&["config", "user.email", "test@test.com"])
            .await
            .unwrap();
        git.run(&["config", "user.name", "Test"]).await.unwrap();

        // Create initial commit
        let file_path = path.join("README.md");
        tokio::fs::write(&file_path, "# Test").await.unwrap();
        git.run(&["add", "."]).await.unwrap();
        git.run(&["commit", "-m", "initial"]).await.unwrap();

        (dir, path)
    }

    #[test]
    fn test_sanitize_branch_normal() {
        assert_eq!(sanitize_branch("main"), "main");
    }

    #[test]
    fn test_sanitize_branch_with_slashes() {
        assert_eq!(sanitize_branch("feature/x"), "feature-x");
    }

    #[test]
    fn test_sanitize_branch_empty() {
        assert_eq!(sanitize_branch(""), "unnamed-branch");
    }

    #[test]
    fn test_sanitize_branch_dot() {
        assert_eq!(sanitize_branch("."), "unnamed-branch");
        assert_eq!(sanitize_branch(".."), "unnamed-branch");
    }

    #[test]
    fn test_sanitize_branch_special_chars() {
        assert_eq!(sanitize_branch("a:b*c"), "a-b-c");
        assert_eq!(sanitize_branch("a?b\"c"), "a-b-c");
        assert_eq!(sanitize_branch("a<b>c|d"), "a-b-c-d");
    }

    #[tokio::test]
    async fn test_create_worktree_for_existing_branch() {
        let (_dir, path) = create_test_repo().await;
        let git = Git::new(&path);
        git.run(&["branch", "feature-test"]).await.unwrap();

        let wm = WorktreeManager::new();
        let wt_path = wm.create("feature-test", &path).await.unwrap();

        assert!(wt_path.exists());

        // Verify worktree is on the correct branch
        let wt_git = Git::new(&wt_path);
        let branch = wt_git.current_branch().await.unwrap();
        assert_eq!(branch, "feature-test");

        // Cleanup
        let _ = wm.remove(&path, &wt_path).await;
    }

    #[tokio::test]
    async fn test_create_skips_main_worktree_check() {
        let (_dir, path) = create_test_repo().await;
        let git = Git::new(&path);

        // The current branch is checked out in the main worktree.
        // After detaching HEAD, create() should succeed because it
        // now skips the main worktree in its duplicate check.
        let current = git.current_branch().await.unwrap();
        git.run(&["branch", "fallback"]).await.unwrap();
        git.run(&["checkout", "fallback"]).await.unwrap();

        let wm = WorktreeManager::new();
        let wt_path = wm.create(&current, &path).await.unwrap();

        assert!(wt_path.exists());

        // Cleanup
        let _ = wm.remove(&path, &wt_path).await;
    }

    #[tokio::test]
    async fn test_create_fails_for_branch_in_existing_worktree() {
        let (_dir, path) = create_test_repo().await;
        let git = Git::new(&path);
        git.run(&["branch", "dup-test"]).await.unwrap();

        let wm = WorktreeManager::new();

        // First creation should succeed
        let wt_path1 = wm.create("dup-test", &path).await.unwrap();

        // Second creation should fail with BranchAlreadyCheckedOut
        let result = wm.create("dup-test", &path).await;
        assert!(result.is_err());
        match result.unwrap_err() {
            GitError::BranchAlreadyCheckedOut { branch, .. } => {
                assert_eq!(branch, "dup-test");
            }
            e => panic!("Expected BranchAlreadyCheckedOut, got {:?}", e),
        }

        // Cleanup
        let _ = wm.remove(&path, &wt_path1).await;
    }

    #[tokio::test]
    async fn test_remove_worktree() {
        let (_dir, path) = create_test_repo().await;
        let git = Git::new(&path);
        git.run(&["branch", "remove-test"]).await.unwrap();

        let wm = WorktreeManager::new();
        let wt_path = wm.create("remove-test", &path).await.unwrap();
        assert!(wt_path.exists());

        wm.remove(&path, &wt_path).await.unwrap();
        // Worktree directory should be gone
        assert!(!wt_path.exists());
    }

    #[tokio::test]
    async fn test_list_managed_excludes_main() {
        let (_dir, path) = create_test_repo().await;
        let git = Git::new(&path);
        git.run(&["branch", "managed-test"]).await.unwrap();

        let wm = WorktreeManager::new();
        let wt_path = wm.create("managed-test", &path).await.unwrap();

        let managed = wm.list_managed(&path).await.unwrap();
        // Should contain only the managed worktree, not the main repo
        assert!(managed.len() >= 1);
        for wt in &managed {
            assert!(!wt.is_main_worktree);
        }

        // Cleanup
        let _ = wm.remove(&path, &wt_path).await;
    }

    #[tokio::test]
    async fn test_list_managed_empty_for_fresh_repo() {
        let (_dir, path) = create_test_repo().await;
        let wm = WorktreeManager::new();

        let managed = wm.list_managed(&path).await.unwrap();
        assert!(managed.is_empty());
    }

    #[tokio::test]
    async fn test_worktree_path_deterministic() {
        let (_dir, path) = create_test_repo().await;
        let wm = WorktreeManager::new();

        let path1 = wm.worktree_path(&path, "feature-x").await;
        let path2 = wm.worktree_path(&path, "feature-x").await;
        assert_eq!(path1, path2);
    }

    #[tokio::test]
    async fn test_worktree_path_differs_for_different_branches() {
        let (_dir, path) = create_test_repo().await;
        let wm = WorktreeManager::new();

        let path1 = wm.worktree_path(&path, "branch-a").await;
        let path2 = wm.worktree_path(&path, "branch-b").await;
        assert_ne!(path1, path2);
    }
}
